config.go

```diff
@@ -0,0 +1,230 @@
+package main
+
+import (
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/ogier/pflag"
+	"github.com/papertrail/remote_syslog2/papertrail"
+	"github.com/papertrail/remote_syslog2/syslog"
+	"github.com/papertrail/remote_syslog2/utils"
+	"gopkg.in/yaml.v2"
+)
+
+const (
+	MIN_REFRESH_INTERVAL = RefreshInterval(10 * time.Second)
+	DEFAULT_CONFIG_FILE  = "/etc/log_files.yml"
+)
+
+type Config struct {
+	ConfigFile      string
+	Files           []string
+	DestHost        string
+	DestPort        int
+	Protocol        string
+	Hostname        string
+	RefreshInterval RefreshInterval
+	ExcludeFiles    RegexCollection
+	ExcludePatterns RegexCollection
+	LogLevels       string
+	DebugLogFile    string
+	PidFile         string
+	UseTCP          bool
+	UseTLS          bool
+	Daemonize       bool
+	Severity        syslog.Priority
+	Facility        syslog.Priority
+	Poll            bool
+	RootCAs         *x509.CertPool
+}
+
+type configfile struct {
+	Files       []string
+	Destination struct {
+		Host     string `yaml:"host"`
+		Port     int    `yaml:"port"`
+		Protocol string `yaml:"protocol"`
+	}
+	Hostname        string          `yaml:"hostname"`
+	RefreshInterval RefreshInterval `yaml:"new_file_check_interval"`
+	ExcludeFiles    RegexCollection `yaml:"exclude_files"`
+	ExcludePatterns RegexCollection `yaml:"exclude_patterns"`
+}
+
+func NewConfig() (*Config, error) {
+	self := &Config{
+		ExcludeFiles:    RegexCollection{},
+		ExcludePatterns: RegexCollection{},
+	}
+	if err := self.load(); err != nil {
+		return nil, err
+	}
+	// parse flags override config file
+	if err := self.override(); err != nil {
+		return nil, err
+	}
+	// check settings and set defaults if needed
+	if err := self.validate(); err != nil {
+		return nil, err
+	}
+	return self, nil
+}
+
+func (self *Config) load() error {
+	log.Infof("Reading configuration file %s", self.ConfigFile)
+	file, err := ioutil.ReadFile(self.ConfigFile)
+	// don't error if the default config file isn't found
+	if os.IsNotExist(err) && self.ConfigFile == DEFAULT_CONFIG_FILE {
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("Could not read the config file: %s", err)
+	}
+	config := &configfile{}
+	if err = yaml.Unmarshal(file, &config); err != nil {
+		return fmt.Errorf("Could not parse the config file: %s", err)
+	}
+	self.Files = config.Files
+	self.DestHost = config.Destination.Host
+	self.DestPort = config.Destination.Port
+	self.Protocol = config.Destination.Protocol
+	self.Hostname = config.Hostname
+	self.RefreshInterval = config.RefreshInterval
+	self.ExcludeFiles = config.ExcludeFiles
+	self.ExcludePatterns = config.ExcludePatterns
+	return nil
+}
+
+func (self *Config) override() error {
+	configfile := pflag.StringP("configfile", "c", DEFAULT_CONFIG_FILE, "Path to config")
```

Comment:
Useful later, perhaps, when we support sending the -HUP signal?

On Wed, Jul 1, 2015 at 10:42 PM Leon Sodhi <notifications@github.com> wrote:

> In config.go
> <https://github.com/papertrail/remote_syslog2/pull/78#discussion_r33722036>
> :
>
> > +	if err = yaml.Unmarshal(file, &config); err != nil {
> > +		return fmt.Errorf("Could not parse the config file: %s", err)
> > +	}
> > +	self.Files = config.Files
> > +	self.DestHost = config.Destination.Host
> > +	self.DestPort = config.Destination.Port
> > +	self.Protocol = config.Destination.Protocol
> > +	self.Hostname = config.Hostname
> > +	self.RefreshInterval = config.RefreshInterval
> > +	self.ExcludeFiles = config.ExcludeFiles
> > +	self.ExcludePatterns = config.ExcludePatterns
> > +	return nil
> > +}
> > +
> > +func (self *Config) override() error {
> > +	configfile := pflag.StringP("configfile", "c", DEFAULT_CONFIG_FILE, "Path to config")
>
> Heh, this is an interesting one. The configfile switch won't do anything
> because we've already loaded the config.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/papertrail/remote_syslog2/pull/78/files#r33722036>.
>